File: src/main/typescript-solutions/validpalindrome/valid-palindrome.ts
/**
 * @param str
 * @returns boolean
 * @description checks if a string is a palindrome using two pointers technique
 */
export const isPalindromeUsingTwoPointers = (str: string): boolean => {
    let left = 0;
    let right = str.length - 1;

    while (left <= right) {
        if (str[left] != str[right]) {
            return false;
        }

        left += 1;
        right -= 1;
    }

    return true;
}

File: src/main/java/validpalindrome/ValidPalindrome.java
package validpalindrome;


/**
 * Note: A palindrome is a word, phrase, or sequence of characters that reads the same backward as forward.
 */
public class ValidPalindrome {
    /**
     * @param str input string
     * @return boolean
     * @implNote Check if a string is a palindrome using the two pointers technique
     */
    public boolean isAPalindromeUsingTwoPointers(String str) {
        int left = 0;
        int right = str.length() - 1;

        while (left <= right) {
            if (str.charAt(left) != str.charAt(right)) {
                return false;
            }
            // if both right and left are equal, we move both pointers inward
            left = left + 1;
            right = right - 1;
        }

        return true;
    }
}
